Burnable Objects & Clothes [Ready]

Ikarrus opened the pull request:
**Let me know if you think I missed something, because there are a LOT of items in this game!**
- Some Items can be set on fire
  - **Not everything burns**. What should or shouldn't burn should work logically. If you think I made an error let me know.
  - Burning items will turn into ashes if not extinguished in time (variable per item)
  - All of a mob's clothing are burned to ashes if the mob is husked by heat
  - Fireproofed clothes will prevent covered clothes from being burned
  - Picking up a burning item will extinguish it, but only if you have heat-proofed gloves
  - Otherwise it just burns it
- Clothes will also have a chance of being shredded if you are caught in a bomb blast
  - Chance of each clothing being destroyed corresponds to severity of blast (90%, 50%, 10%)
  - Outer layer clothing (headgear and exosuits) take the shock first, which absorbs some or all of the shock for the clothes it covers, according to it's bomb 
- Objects destroyed in these ways will drop its contents. 
- Numbers subject to change
##### TODO:
- [x] ~~Maybe expand functionality to non-human mobs~~
- [x] Maybe expand this to include all items. If I am feeling brave. 
- [x] Maybe expand to all objects, so wooden tables and stuff can burn, too.
- [x] ~~Investigate how acid works and maybe port that into this feature~~
- [x] Add fireproofing to clothing that shouldn't be burnable
- [x] Review bomb armor values to make sure they're appropriate
- [x] Review code for optimization, bugfixing

Ikarrus committed:
Destructible Clothing [WIP]

- Clothes can be set on fire and extinguished
  - Any item can technically be set on fire and extinguished now, but only clothes (and paper) actually burn right now
  - All of a mob's clothing are burned if the mob is husked by fire
- Clothes will also have a chance of being shredded if you are caught in a bomb blast
  - The bigger the explosion, the higher the chance they have of being destroyed
  - Outer layer clothing (headgear and exosuits) take the shock first, which absorbs some of the shock for the clothes it covers, according to it's bomb armor

Ikarrus committed:
Squashed Commits

RemieRichards commented on code/game/objects/items.dm:
```diff
@@ -102,6 +103,18 @@
+/obj/item/proc/extinguish()
+	if(burning)
+		burning = 0
+		for(var/image/img in overlays)
```
PLEASE NO.

store a reference to the fire image (I assume there is only one) on the item, and simply remove it then null it.
bonus points if the fire image is generic enough and you define it as a global/static so only one image exists ever (if there is only one generic image you don't need to reference it on the item, since the global/static var itself is the reference)

RemieRichards commented on code/modules/clothing/clothing.dm:
```diff
@@ -17,6 +17,31 @@
+	src.visible_message("<span class='warning'>[src] catches on fire.</span>")
+	burning = 1
+	overlays += image("icon" = 'icons/effects/fire.dmi', "icon_state" = "fire")
+	spawn(100)
```
don't use a spawn for the cooldown, use ```world.time``` and a cooldown define eg #define ITEM_BURN_COOLDOWN
```
if(world.time >= burntime)
    burntime = world.time + ITEM_BURN_COOLDOWN
    do_the_burning
```

Ikarrus replied:
How could I do this without having to have every item use process()?

A lot of items have their own process() proc, too, so I don't think I can even use that without complications.

AlexCrimson commented:
Kewl. Now give botany the ability to make new jumpsuits.

WJohn commented:
I think burning clothing should be based on the whole suit layer system that acid and now explosive go through as well as strip time (riot suits would take significantly longer to burn than jumpsuits).

Incoming5643 commented:
Unforeseen conquence: Someone will use this to erp by getting pumped full of kelotane then setting themselves on fire in front of prospective stutter nurses.

AlexCrimson commented:
Taking your clothes off is for casuals. Hardcore players will just carry combustion pills.

phil235 commented on code/modules/mob/living/carbon/human/inventory.dm:
```diff
@@ -390,3 +390,85 @@
+
+	if(shredded)
+		if(Human) //Unequip if equipped
+			Human.unEquip(src)
```
What happens if the item has NODROP?

Ikarrus replied:
Good question. I will investigate.

Ikarrus replied:
It looks like the human unequip proc forces them to drop it regardless of flags

phil235 commented:
Adding acid stuff to that system would be cool.

Ikarrus commented:
Acid seems to work differently in that it doesn't start dealing damage to the mob until the clothes on top are melted, and only on a per-limb basis.

I feel kind of bad about having some megaproc that has a bunch of extra vars and rules just to lump acid into it.

Lumipharon commented:
Burning people's jumpsuits so all the shit falls out of their pockets/id/belt etc sounds pretty awful.
You'd have to be pretty heavily on fire to burn up in 10 seconds.

Ikarrus commented:
Clothing you're wearing only burns up when you get husked from heat. 

Sent from my iPhone

> On Jun 25, 2015, at 4:25 PM, Lumipharon <notifications@github.com> wrote:
> 
> Burning people's jumpsuits so all the shit falls out of their pockets/id/belt etc sounds pretty awful.
> You'd have to be pretty heavily on fire to burn up in 10 seconds.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Lumipharon commented:
I dunno if you want to let people/AI's plasma flood the station and psuedo acid foam EVERYTHING.
But that's just my opinion.

Ikarrus commented:
The portion of burnable objects in the game would probably be considerably smaller than you'd think.

Ikarrus commented:
Items can burn now!!

Ikarrus committed:
Resolved Conflicts:
	code/modules/clothing/clothing.dm
	code/modules/clothing/masks/gasmask.dm
	code/modules/clothing/suits/bio.dm
	code/modules/clothing/suits/utility.dm

Ikarrus commented:
PR is done and ready to face the full brunt of your review attacks.

Razharas commented:
So if you ignite something of fire, then extinguish, then wait burntime-1 and ignite again it will burn?
I cant say i like it this way

Ikarrus committed:
Burning objects are processed every tick now instead of using a spawn() check

Ikarrus commented:
>So if you ignite something of fire, then extinguish, then wait burntime-1 and ignite again it will burn?
I cant say i like it this way

Fixed

Steelpoint commented:
>clothing being destroyed by bombs.

This is a bad change, a shame the title of this PR gave no indication that it also made bombs remove clothing again.

Ikarrus commented:
What exactly makes it a bad change for you?

Steelpoint commented:
One quick example I can list is losing my Wizard robes to a small explosion.

I can count more, the fact of the matter is this feature was added a while ago and it was heavily disliked enough until it was reverted.

Explosions removing clothing or any other items on one's person has never been a good feature.

Ikarrus commented:
Your information must be outdated then, because shredding have been removed from light explosions since.

It should only shred them if the explosion was powerful enough to put you into crit.

Steelpoint commented:
It was a explosive round, it did not put my into crit but I lost my robes and thus was defenceless. Fortunately this was not a round start Wizard but a admin created one.

Nevertheless, removal and deletion of items (clothing or otherwise) from explosions is a bad game play mechanic in my opinion, its punishing enough to get a bomb thrown in your face and all this PR did was further punish you for getting a bomb thrown at you by removing your clothing and other items from the game entirely.

Ikarrus commented:
It should only be shredding clothes if you would normally be put into crit. If that is not the case, then either the server was not up to date at the time or it's a bug.

Steelpoint commented:
it was a day ago, I know for certain I was hit by a single explosive slug from a syndicate shotgun wielded by a security officer.

Ikarrus commented:
Make an issue report please

Sent from my iPhone

> On Jul 11, 2015, at 9:35 AM, Steelpoint <notifications@github.com> wrote:
> 
> it was a day ago, I know for certain I was hit by a single explosive slug from a syndicate shotgun wielded by a security officer.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Steelpoint commented:
Attack log from public logs

[11:14:26]ATTACK: Happi Futur(kingofcarrotflowerspt1) shot Frank the Fabulous (as Sofo Oofs)(steelpoint) with The bullet (NEWHP: 4)
